crates/ansi-to-html/src/ansi/parse.rs

```diff
@@ -0,0 +1,216 @@
+use std::{ops::Range, str::CharIndices};
+
+#[must_use]
+pub(crate) struct AnsiParser<'text> {
+    text: &'text str,
+    chars: CharIndices<'text>,
+}
+
+impl<'text> AnsiParser<'text> {
+    pub(crate) fn new(text: &'text str) -> Self {
+        let chars = text.char_indices();
+        Self { text, chars }
+    }
+}
+
+#[derive(Debug)]
+pub(crate) enum AnsiFragment<'text> {
+    Sequence(&'text str),
+    Text(&'text str),
+}
+
+impl<'text> Iterator for AnsiParser<'text> {
+    type Item = AnsiFragment<'text>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut dfa = Dfa::new(&mut self.chars).unwrap();
+
+        match dfa.next()? {
+            // Walk through the current run of text
+            Status::Waiting => {
+                while dfa.peek() == Some(Status::Waiting) {
+                    dfa.next();
+                }
+                Some(AnsiFragment::Text(&self.text[dfa.span()]))
+            }
```

Comment:
We could actually avoid peeking here entirely as the only character that breaks us out of this loop is escape, so we can eagerly pop off the escape in here and account for that the next time that we start iterating

I think we could also avoid peeking in the `Status::InSequence` branch too through more significant changes that allow for walking back the iterator on an invalid character

Reply:
Thinking on this more there's definitely a lot of optimizations that can be done (that'd I'd prefer doing in follow-up PRs)

I wish some kind of generator was already stable :sweat:

Reply:
Thinking on this more. Both cases can be generalized to adding support for storing the character that invalidated the current run of text / ansi sequences and `.take()`ing that into account the next time you start iterating